index.html

```diff
@@ -3102,22 +3116,44 @@ <h3>Verify Proof Sets and Chains</h3>
 Set <var>allProofs</var> to </var><var>securedDocument</var>.<var>proof</var>.
           </li>
           <li>
-For each <var>proof</var> in <var>allProofs</var>, run steps 2 through 11 of the algorithm in
+For each <var>proof</var> in <var>allProofs</var>, do the following steps:
+          </li>
+          <ol class="algorithm">
+            <li>
+Let <var>matchingProofs</var> be an empty list.
+            </li>
+            <li>
+If <var>proof</var> contains a <code>previousProof</code> attribute and that
+attribute is a string, add the element from <var>allProofs</var> with an
+<code>id</code> attribute matching <code>previousProof</code> to matchingProofs.
+If a proof with <code>id</code> does not exist in <var>allProofs</var>, an error
+MUST be raised and SHOULD convey an error type of
+<a href="#MALFORMED_PROOF_ERROR">MALFORMED_PROOF_ERROR</a>.
+If the <code>previousProof</code> attribute is an array, add each element from
+<var>allProofs</var> with an <code>id</code> attribute that matches an element
+of that array. If any matches are missing, an error MUST be raised and SHOULD
+convey an error type of
+<a href="#MALFORMED_PROOF_ERROR">MALFORMED_PROOF_ERROR</a>.
+            </li>
+            <li>
```

Comment:
See my comment below about needing to keep some text about verifying previous proofs before reading more here.

Now, once we have the `matchingProofs` here, we need to check that each of them has been verified already. I recommend we express this somewhat simply as we did before (see text currently marked to be removed below near my other comment) and allow implementers room to decide how to best ensure each previous proof was already verified (e.g., recursively, with a non-recursive scanning algorithm, etc.), but state that if any cycles are detected the proof cycle error must be thrown as before.

If we wanted to be prescriptive, one way to do this would be to use a map to populate with keys that are matching proof IDs and values that keep track of whether the proof is currently being verified or has been verified (absence in that map means the proof hasn't been seen at all yet). This map would be passed along / shared during verification of all related proofs, being initialized only at the outset (if it wasn't passed in as an option). We would then iterate through each matching proof and, if the proof ID was already in the map and not verified, we would throw the cycle error, otherwise we would add it and recurse into this algorithm, passing the map. Each proof, upon verification, would have its entry updated to indicate as such -- and a verification failure would terminate the whole algorithm.

But, again, I think it's better to allow implementations to do the above part however they want to.

Reply:
Actually, ignore this comment -- I see that a different, simpler change is possible, I'll do another review.